doc/user_guide/introduction_to_psykal.rst

```diff
@@ -0,0 +1,408 @@
+(``matrix_vector_kernel_mm_type``).
+The ``setval_c`` Built-in sets all values in the field ``Ax`` to ``1.0`` and
+the ``X_divideby_Y`` Built-in divides values in the field ``rhs`` by their
+equivalent (per degree of freedom) values in the field ``lumped_weight``
+(see :ref:`supported LFRic API Built-ins <lfric-built-ins>`).
+Notice that, unlike the kernel call, no ``use`` association is required for
+the Built-ins since they are provided as part of the environment.
+::
+
+  module solver_mod
+    ...
+    use matrix_vector_mm_mod, only: matrix_vector_kernel_mm_type
+    ...
+
+    subroutine jacobi_solver_algorithm(lhs, rhs, mm, mesh, n_iter)
+
+      integer(kind=i_def), intent(in)    :: n_iter
+      type(field_type),    intent(inout) :: lhs
+      type(field_type),    intent(in)    :: rhs
+      type(operator_type), intent(in)    :: mm
+      type(mesh_type),     intent(in)    :: mesh
+      type(field_type)                   :: Ax, lumped_weight, res
+
+      real(kind=r_def), parameter :: MU = 0.9_r_def
+      ...
+
+      ! Compute mass lump
+      call invoke( name = "Jacobi_mass_lump",                           &
+                   setval_c(Ax, 1.0_r_def),                             &
+                   matrix_vector_kernel_mm_type(lumped_weight, Ax, mm), &
+                   X_divideby_Y(lhs, rhs, lumped_weight) )
+
+    end subroutine jacobi_solver_algorithm
+    ...
+
+  end module solver_mod
+
+
+.. _PSy-layer:
+
+PSy layer
+---------
+
+In the PSyKAl model, the PSy layer is the bridge between the Algorithm
+full field operations and the Kernel and Built-Ins individual element
+operations. As such, it is responsible for:
+
+1. calling any Kernel and expanding any Buit-In so that they iterate over
+   their specified interation space;
+2. map the Kernel and Built-In arguments supplied by an Algorithm ``invoke``
+   call to the arguments required by a Built-in or Kernel methods;
+3. include any required distributed-memory operations such as halo swaps
+   and reductions to guarantee the correctness of the code;
+4. and enable an entry point for the optimisation expert to provide
+   additional optimisations for the target architecture.
+
+The PSy layer can be written manually but this is error prone and
+potentially complex to optimise. Therefore, the PSyclone code generation
+system automatically generates an initial version of the PSy layer by
+parsing the associated Algorithm file and each of the kernels used on it.
+This initial version of the PSy-layer can be further tuned with the PSyclone
+code transformation capabilities by providing a transformation script.
+Each API comes with a set of specialise transformations designed for that API.
+
+Also, in addition to all the functionality available for PSyIR Nodes, the
+PSy-layer nodes have a `dag()` method (standing for directed acyclic graph),
+outputs the PSyIR nodes and its data dependencies. By default a file in
```

Comment:
"which outputs...and their data dependencies."

Reply:
Done